### PR TITLE
Add cosine_similarity Presto function

### PR DESCRIPTION
**Pull request by amitkdutta**
Summary:
- https://en.wikipedia.org/wiki/Cosine_similarity
- [Presto Java implementation](https://github.com/prestodb/presto/blob/66aab8b7ef886fcebf38fae2d2e9094f12fe3f6b/presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java#L1627)

Differential Revision: D50589652

### CODE REVIEW

File: velox/docs/functions/presto/math.rst [hunk 1]
```diff
@@ -27,6 +27,15 @@ Mathematical Functions
     verified for performance reasons. Returns ``high`` for all values of ``x``
     when ``low`` is greater than ``high``.
 
+.. function:: cosine_similarity(map(varchar, double), map(varchar, double)) -> double
+
+    Returns the cosine similarity between the sparse vectors x and y. If any value of any key is NULL,
```
Comment (mbasmanova) [inline velox/docs/functions/presto/math.rst:32]: perhaps, turn "cosine similarity" into a link to https://en.wikipedia.org/wiki/Cosine_similarity to provide convenient access to unfamiliar readers

File: velox/docs/functions/presto/math.rst [hunk 2]
```diff
@@ -27,6 +27,15 @@ Mathematical Functions
...
+    the function returns NULL.
+
+    For example:
```
Comment (mbasmanova) [inline velox/docs/functions/presto/math.rst:35]: The formatting here might be off. Please, generate docs locally and verify. Would be great to share a screenshot.

File: velox/functions/prestosql/Arithmetic.h [hunk 3]
```diff
@@ -531,5 +532,42 @@ struct WilsonIntervalLowerFunction {
...
+      }
+    }
+    return result;
+  }
+
+  bool callNullFree(
+      out_type<double>& result,
+      const null_free_arg_type<Map<Varchar, double>>& leftMap,
+      const null_free_arg_type<Map<Varchar, double>>& rightMap) {
+    double normLeftMap = normalizeMap(leftMap);
```
Comment (mbasmanova) [inline velox/functions/prestosql/Arithmetic.h:564]: nit: normalizedLeft and normalizedRight

File: velox/functions/prestosql/Arithmetic.h [hunk 4]
```diff
@@ -531,5 +532,42 @@ struct WilsonIntervalLowerFunction {
...
+    for (const auto& [key, value] : leftMap) {
+      auto it = rightMap.find(key);
+      if (it != rightMap.end()) {
+        result += value * it->second;
+      }
+    }
+    return result;
+  }
+
+  bool callNullFree(
```
Comment (mbasmanova) [inline velox/functions/prestosql/Arithmetic.h:560]: Can this function be void?

File: velox/functions/prestosql/tests/ArithmeticTest.cpp [hunk 5]
```diff
@@ -821,5 +821,44 @@ TEST_F(ArithmeticTest, wilsonIntervalUpper) {
   EXPECT_DOUBLE_EQ(wilsonIntervalUpper(1, 3, kInf).value(), 1.0);
 }
 
+TEST_F(ArithmeticTest, cosineSimilarity) {
+  const auto testCosineSimilarity =
```
Comment (mbasmanova) [inline velox/functions/prestosql/tests/ArithmeticTest.cpp:825]: In other places we do not add tests prefix to such lambdas

```
const auto cosineSimilarity =
```

File: velox/functions/prestosql/tests/ArithmeticTest.cpp [hunk 6]
```diff
@@ -821,5 +821,44 @@ TEST_F(ArithmeticTest, wilsonIntervalUpper) {
...
+      [&](const std::vector<
+              std::vector<std::pair<std::string, std::optional<double>>>>& left,
+          const std::vector<
+              std::vector<std::pair<std::string, std::optional<double>>>>&
+              right,
+          double expected) {
+        auto leftMap = makeMapVector<std::string, double>(left);
+        auto rightMap = makeMapVector<std::string, double>(right);
+        EXPECT_DOUBLE_EQ(
```
Comment (mbasmanova) [inline velox/functions/prestosql/tests/ArithmeticTest.cpp:834]: Move EXPECT_DOUBLE_EQ to the caller to provide more useful error messages. Have this function return the result instead.

```
EXPECT_DOUBLE_EQ(
   (2.0 * 3.0) / (std::sqrt(5.0) * std::sqrt(10.0)),
   cosineSimilarity({{{"a", 1}, {"b", 2}}},...)
)
```

File: velox/functions/prestosql/tests/ArithmeticTest.cpp [hunk 7]
```diff
@@ -821,5 +821,44 @@ TEST_F(ArithmeticTest, wilsonIntervalUpper) {
   EXPECT_DOUBLE_EQ(wilsonIntervalUpper(1, 3, kInf).value(), 1.0);
 }
 
+TEST_F(ArithmeticTest, cosineSimilarity) {
+  const auto testCosineSimilarity =
+      [&](const std::vector<
```
Comment (mbasmanova) [inline velox/functions/prestosql/tests/ArithmeticTest.cpp:826]: Why do we take a list of maps? This is a simple function, hence, no need to test it on multi-row vectors. Let's simplify the test.

**Review by mbasmanova — COMMENTED** [review]
Looks nice % some minor comments.

**Comment by amitkdutta** [comment]
Screenshot from 2023-11-03 17-49-32

Tested doc locally and made cosine similarity a URL pointing to wiki.

**Comment by amitkdutta** [comment]
Fixed documentation indentation for SQL

[
Screenshot from 2023-11-03 22-03-20
](url)

File: velox/docs/functions/presto/math.rst [hunk 8]
```diff
@@ -27,6 +27,17 @@ Mathematical Functions
     verified for performance reasons. Returns ``high`` for all values of ``x``
     when ``low`` is greater than ``high``.
 
+.. function:: cosine_similarity(map(varchar, double), map(varchar, double)) -> double
+
+    Returns the `cosine similarity <https://en.wikipedia.org/wiki/Cosine_similarity>`_ between the vectors represented as map(varchar, double). If any value
+    of any key is NULL, the function returns NULL.
+
+    Example::
```
Comment (mbasmanova) [inline velox/docs/functions/presto/math.rst:35]: nit: drop "Example". It takes up space without adding value.

File: velox/expression/ConjunctExpr.h [hunk 9]
```diff
@@ -82,10 +82,6 @@ class ConjunctExpr : public SpecialForm {
       FlatVector<bool>* result,
       SelectivityVector* activeRows);
 
-  bool evaluatesArgumentsOnNonIncreasingSelection() const override {
```
Comment (mbasmanova) [inline velox/expression/ConjunctExpr.h:85]: Unrelated changes?

File: velox/docs/functions/presto/math.rst [hunk 10]
```diff
@@ -27,6 +27,17 @@ Mathematical Functions
     verified for performance reasons. Returns ``high`` for all values of ``x``
     when ``low`` is greater than ``high``.
 
+.. function:: cosine_similarity(map(varchar, double), map(varchar, double)) -> double
```
Comment (mbasmanova) [inline velox/docs/functions/presto/math.rst:30]: Would you document what happens if one or both input maps are empty?

**Review by mbasmanova — COMMENTED** [review]
Overall looks good % there are unrelated change, some questions and a question about how much fuzzer coverage we get for this function.

**Comment by mbasmanova** [comment]
@kgpai Krishna, there is a CI failure in Build and test pyVelox. Could you help us understand what it means?

```
======================================================================
FAIL: test_function_signature (pyvelox.test.test_signatures.TestFunctionSignatures)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/root/project/pyvelox/test/test_signatures.py", line 44, in test_function_signature
    self.assertEqual(str(concat_signatures[0].return_type()), "array(__user_T1)")
AssertionError: 'array(date)' != 'array(__user_T1)'
- array(date)
+ array(__user_T1)
```

**Comment by kgpai** [comment]
@mbasmanova @amitkdutta This is a test case failure , the test is just checking that concat function has right return type . This test is just to ensure that we can pick up the right signatures and uses concat as its very common. Amit’s changes shouldnt have anything to do with that failure, I am rerunning it and if it still fails I will fix it.

**Comment by amitkdutta** [comment]
@kgpai The failures show fuzzer and pyvelox errors.
Screenshot from 2023-11-07 09-21-28
Its also confusing when I open `ci/circleci: linux-pr-fuzzer-run` - it shows pyvelox errors. Does rebasing fix it? Also why we don't see this pyvelox in meta internal tests? Is there a place where I can see the lifecycle of this tests (e.g. is it failing for some time or just this PR tirggered it)

File: velox/docs/functions/presto/math.rst [hunk 11]
```diff
@@ -27,6 +27,17 @@ Mathematical Functions
...
+
+    Returns the `cosine similarity <https://en.wikipedia.org/wiki/Cosine_similarity>`_ between the vectors represented as map(varchar, double). If any value
```
Comment (mbasmanova) [inline velox/docs/functions/presto/math.rst:32]: > If any value of any key is NULL, the function returns NULL.

This statement is redundant as this is the default for all Presto functions.

File: velox/docs/functions/presto/math.rst [hunk 12]
```diff
@@ -27,6 +27,17 @@ Mathematical Functions
...
+    of any key is NULL, the function returns NULL. If any input map is empty, the function returns NaN.
+
+        SELECT cosine_similarity(MAP(ARRAY['a'], ARRAY[1.0]), MAP(ARRAY['a'], ARRAY[2.0])); -- 1.0
+
+        SELECT cosine_similarity(MAP(ARRAY['a', 'b'], ARRAY[1.0, 2.0]), MAP(ARRAY['a', 'b'], ARRAY[NULL, 3.0])); -- NULL
+
+        SELECT cosine_similarity(MAP(ARRAY[], ARRAY[]), MAP(ARRAY[], ARRAY[])); -- NaN
```
Comment (mbasmanova) [inline velox/docs/functions/presto/math.rst:39]: Perhaps, make one of the maps non-empty.

**Review by mbasmanova — APPROVED** [review]
Thanks. Makes sure CI is green before merging.

File: velox/functions/prestosql/tests/ArithmeticTest.cpp [hunk 13]
```diff
@@ -821,5 +821,46 @@ TEST_F(ArithmeticTest, wilsonIntervalUpper) {
...
+
+  EXPECT_DOUBLE_EQ(
+      (2.0 * 3.0 + (-1) * 1) / (std::sqrt(1 + 4 + 1) * std::sqrt(1 + 9)),
+      cosineSimilarity({{"a", 1}, {"b", 2}, {"c", -1}}, {{"c", 1}, {"b", 3}}));
+
+  EXPECT_DOUBLE_EQ(
+      0.0,
+      cosineSimilarity({{"a", 1}, {"b", 2}, {"c", -1}}, {{"d", 1}, {"e", 3}}));
+
+  auto nullableLeftMap = makeNullableMapVector<StringView, double>(
```
[the remark this answers is not in the document]
Comment (amitkdutta) [reply velox/functions/prestosql/tests/ArithmeticTest.cpp:859]: Done

File: velox/functions/prestosql/Arithmetic.h [hunk 14]
```diff
@@ -531,5 +532,41 @@ struct WilsonIntervalLowerFunction {
...
+    }
+    return result;
+  }
+
+  void callNullFree(
+      out_type<double>& result,
+      const null_free_arg_type<Map<Varchar, double>>& leftMap,
+      const null_free_arg_type<Map<Varchar, double>>& rightMap) {
+    double normLeftMap = normalizeMap(leftMap);
+    double normRightMap = normalizeMap(rightMap);
```
[the remark this answers is not in the document]
Comment (amitkdutta) [reply velox/functions/prestosql/Arithmetic.h:575]: Added check and returned NAN.

File: velox/docs/functions/presto/math.rst (hunk 10 above)
Comment (amitkdutta) [reply velox/docs/functions/presto/math.rst:30]: Done

File: velox/docs/functions/presto/math.rst (hunk 2 above)
Comment (mbasmanova) [reply velox/docs/functions/presto/math.rst]: That would be nice. CC: @kgpai
